src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,163 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic {
+  namespace util {
+
+    template<class T> 
+    class SlidingWindow {
+      public:
+        SlidingWindow(UInt maxCapacity);
+        SlidingWindow(UInt maxCapacity, std::vector<T> initialData);
+        const UInt maxCapacity;
+        const T NEUTRAL_VALUE = 0.0;
+        UInt size() const;
+        /** append new value to the end of the buffer and handle the "overflows"-may pop the first element if full. 
+         :return addition(+) neutral value (that is 0) when size()< maxCapacity; when full - return the value of the dropped element  
+        */
+        T append(T newValue);
+        /**
+        * :return unordered content (data ) of this sl. window; call linearize() if you need them oredered from oldest->newest
+        * This direct access method is fast.
+        */
+        const std::vector<T>& getData() const;
+
+        /** linearize method for the internal buffer; this is slower than the pure getData() but ensures that the data are ordered (oldest at the 
+         * beginning, newest at the end of the vector
+         * This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+         * :return new linearized vector
+        */
+        std::vector<T> linearize() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+
+    // HELPER:
+     static std::vector<T> getLastNValues(std::vector<T> biggerData, nupic::UInt n) {
+      NTA_CHECK(n >=0 && n <= biggerData.size()); 
+    std::vector<T> v;
+    v.reserve(n);
+    v.insert(begin(v), end(biggerData) - n, end(biggerData));
+    NTA_ASSERT(v.size() == n);
+    return v;
+    };
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+        bool firstRun_ = true; //fill be false once "first run" (=size reaches maxCapacity for the first time) is completed
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::util;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  idxNext_ = 0;
+}
+
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(UInt maxCapacity, vector<T> initialData)  :
+  SlidingWindow(maxCapacity) {
+  NTA_CHECK(initialData.size() <= maxCapacity);
```

Comment:
>  it would be reasonable to just loop over the array and call append on each element. 

Wouldn't we waste on a size-3 sliding window with initialization of 1000 items? Or I'm just overthinking it?